src/dbnode/storage/index.go

```diff
@@ -1643,6 +1701,48 @@ func (i *nsIndex) execBlockQueryFn(
 	state.exhaustive = state.exhaustive && blockExhaustive
 }
 
+func (i *nsIndex) execBlockWideQueryFn(
+	ctx context.Context,
+	cancellable *resource.CancellableLifetime,
+	block index.Block,
+	query index.Query,
+	opts index.QueryOptions,
+	state *asyncQueryExecState,
+	results index.BaseResults,
+	logFields []opentracinglog.Field,
+) {
+	logFields = append(logFields,
+		xopentracing.Time("blockStart", block.StartTime()),
+		xopentracing.Time("blockEnd", block.EndTime()),
+	)
+
+	ctx, sp := ctx.StartTraceSpan(tracepoint.NSIdxBlockQuery)
+	sp.LogFields(logFields...)
+	defer sp.Finish()
+
+	_, err := block.Query(ctx, cancellable, query, opts, results, logFields)
+	if err == index.ErrUnableToQueryBlockClosed {
+		// NB(r): Because we query this block outside of the results lock, it's
+		// possible this block may get closed if it slides out of retention, in
+		// that case those results are no longer considered valid and outside of
+		// retention regardless, so this is a non-issue.
+		err = nil
+	} else if err == index.ErrWideQueryResultsExhausted {
+		// NB: this error indicates a wide query short-circuit, so it is expected
+		// after the queried shard set is exhausted.
+		err = nil
+	}
+
+	state.Lock()
+	defer state.Unlock()
+
+	if err != nil {
+		sp.LogFields(opentracinglog.Error(err))
+		state.multiErr = state.multiErr.Add(err)
+	}
+	state.exhaustive = true
```

Comment:
Shouldn't this be `state.exhaustive = state.exhaustive && blockExhaustive` where `blockExhaustive` is returned from block.Query?

Reply:
Since this is performing a wide quer, it's always exhaustive so we should be ok here :+1:

Reply:
Sounds good, resolving comment.